dm/pkg/binlog/pos_finder.go

```diff
@@ -0,0 +1,369 @@
+	parser.SetVerifyChecksum(true)
+
+	return &binlogPosFinder{
+		remote:     false,
+		tctx:       tctx,
+		enableGTID: enableGTID,
+		parser:     parser,
+		flavor:     flavor,
+
+		relayDir: relayDir,
+	}
+}
+
+func NewRemoteBinlogPosFinder(tctx *tcontext.Context, db *sql.DB, syncCfg replication.BinlogSyncerConfig, enableGTID bool) *binlogPosFinder {
+	// make sure raw mode enabled, and MaxReconnectAttempts set
+	syncCfg.RawModeEnabled = true
+	if syncCfg.MaxReconnectAttempts == 0 {
+		syncCfg.MaxReconnectAttempts = common.MaxBinlogSyncerReconnect
+	}
+
+	parser := replication.NewBinlogParser()
+	parser.SetFlavor(syncCfg.Flavor)
+	parser.SetVerifyChecksum(true)
+
+	return &binlogPosFinder{
+		remote:     true,
+		tctx:       tctx,
+		enableGTID: enableGTID,
+		parser:     parser,
+		flavor:     syncCfg.Flavor,
+
+		db:      db,
+		syncCfg: syncCfg,
+	}
+}
+
+func (r *binlogPosFinder) getBinlogFiles() (FileSizes, error) {
+	if r.remote {
+		return GetBinaryLogs(r.tctx.Ctx, r.db)
+	} else {
+		return GetLocalBinaryLogs(r.relayDir)
+	}
+}
+
+func (r *binlogPosFinder) startSync(position mysql.Position) (reader.Reader, error) {
+	if r.remote {
+		binlogReader := reader.NewTCPReader(r.syncCfg)
+		return binlogReader, binlogReader.StartSyncByPos(position)
+	} else {
+		binlogReader := reader.NewFileReader(&reader.FileReaderConfig{EnableRawMode: true})
+		position.Name = path.Join(r.relayDir, position.Name)
+		return binlogReader, binlogReader.StartSyncByPos(position)
+	}
+}
+
+func (r *binlogPosFinder) findMinTimestampOfBinlog(currBinlog binlogSize) (uint32, error) {
+	var minTs uint32
+	binlogReader, err := r.startSync(mysql.Position{Name: currBinlog.name, Pos: FileHeaderLen})
+	if err != nil {
+		return 0, err
+	}
+	for {
+		ev, err := binlogReader.GetEvent(r.tctx.Ctx)
+		if err != nil {
+			binlogReader.Close()
+			return 0, err
+		}
+		// break on first non-fake event(must be a format description event)
+		if !utils.IsFakeRotateEvent(ev.Header) {
+			minTs = ev.Header.Timestamp
+			break
+		}
+	}
+	binlogReader.Close()
+
+	return minTs, nil
+}
+
+func (r *binlogPosFinder) initTargetBinlogFile(ts int64) error {
+	targetTs := uint32(ts)
+	var lastTs, minTs uint32
+	var lastMid int
+	binaryLogs, err := r.getBinlogFiles()
+	if err != nil {
+		return err
+	}
+	if len(binaryLogs) <= 0 {
+		// should not happen on a master with binlog enabled
+		return errors.New("cannot find binlog files")
+	}
+
+	begin, end := 0, len(binaryLogs)-1
+	for begin <= end {
+		mid := (begin + end) / 2
+		currBinlog := binaryLogs[mid]
+
+		minTs, err = r.findMinTimestampOfBinlog(currBinlog)
+		if err != nil {
+			return err
+		}
+
+		r.tctx.L().Debug("min timestamp in binlog file", zap.Reflect("file", currBinlog), zap.Uint32("ts", minTs))
+
+		lastTs = minTs
+		lastMid = mid
+
+		if minTs >= targetTs {
+			end = mid - 1
+		} else {
+			// current binlog maybe the target binlog file, we'll backtrace to it later.
+			begin = mid + 1
+		}
+	}
+	if lastTs >= targetTs {
```

Comment:
is `lastTs >= targetTs` means we can't find user wanted binlog file?, do we still need to fill targetBinlog field?

Reply:
2 cases may causes lastTs >= targetTs, target timestamp either below smallest timestamp in binlog files in that case we need to return the begin of the first binlog and use `posType` to indicate or we need backtrace to previous file.

Reply:
got this, we need to return min binlog file and gtid set in  error msg in this case

Reply:
in the first case, we need to return the min position without error. When user start a task for the first time, may give a timestamp much smaller than min binlog timestamp.